File: Difficulty: Medium/LCS of three strings/lcs-of-three-strings.py
class Solution:
    def lcsOf3(self,s1, s2, s3):
        # Code here
        n1, n2, n3 = len(s1), len(s2), len(s3)
    
        
        dp = [[[0]*(n3+1) for _ in range(n2+1)] for _ in range(n1+1)]
        for i in range(n1):
            for j in range(n2):
                for k in range(n3):
                    if s1[i] == s2[j] == s3[k]:
                        dp[i+1][j+1][k+1] = dp[i][j][k] + 1
                    else:
                        dp[i+1][j+1][k+1] = max(dp[i+1][j+1][k], dp[i][j+1][k+1], dp[i+1][j][k+1])
        
        return dp[n1][n2][n3]

File: Difficulty: Medium/LCS of three strings/lcs-of-three-strings.java
class Solution {
    String a,b,c;
    int l,m,n;
    public int help(int i,int j,int k,int dp[][][]){
        if(i==l || j==m ||k==n){
            return 0;
        }
        if(dp[i][j][k]!=-1)return dp[i][j][k];
        if(a.charAt(i)==b.charAt(j) && a.charAt(i)==c.charAt(k)){
            return dp[i][j][k]=1+help(i+1,j+1,k+1,dp);
        }
        return dp[i][j][k]=Math.max(
                    help(i+1, j, k,dp),
                    Math.max(
                        help(i, j+1, k,dp),
                        Math.max(
                            help(i, j, k+1,dp),
                            Math.max(
                                help(i+1, j+1, k,dp),
                                Math.max(
                                    help(i+1, j, k+1,dp),
                                    help(i, j+1, k+1,dp)
                                )
                            )
                        )
                    )
                );
    }

    int lcsOf3(String s1, String s2, String s3) {
        // code here
        a=s1;
        b=s2;
        c=s3;
        l = s1.length();
        m = s2.length();
        n = s3.length();
        int dp[][][]=new int[l+1][m+1][n+1];
        for(int i=0;i<l;i++){
            for(int j=0;j<m;j++){
                Arrays.fill(dp[i][j],-1);
            }
        }
        
        return help(0,0,0,dp);
    }
}